src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1452,6 +1452,11 @@ private static LambdaConversionResult IsAnonymousFunctionCompatibleWithDelegate(
                         {
                             return LambdaConversionResult.MismatchedParameterDefaultValue;
                         }
+
+                        if (lambdaSymbol.Parameters[p].IsParams != delegateParameters[p].IsParams)
+                        {
+                            return LambdaConversionResult.MismatchedParamsArray;
+                        }
```

Comment:
Spec question: why do we want `params` differences to result in no conversion? The same question applies to optional parameters.

Reply:
Also, do we need a spec update for anonymous function conversions (section 10.7, [https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/conversions.md#107-anonymous-function-conversions](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/conversions.md#107-anonymous-function-conversions) )? 
I don't see the term mentioned in the spec [https://github.com/dotnet/csharplang/blob/main/proposals/lambda-method-group-defaults.md](https://github.com/dotnet/csharplang/blob/main/proposals/lambda-method-group-defaults.md)

Reply:
Tagging @captainsafia for question above (why should `params` differences result in no conversion?)

Reply:
The following method group conversion is allowed today ([sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKAwGYACbUgYVMIG9DTHSATAUxVYHMIFXyAWUgBEAFAAcIAJwhgAzqVgIA2gF0FASgDcDJjsYkBpALIjFqjTQJNS9K9aZCWpALzHtdxgF89lz0A===)):
```
    delegate void D(params int[] i);
    
    public void M(int[] i) 
    {
        D d = M;
    }
```

Reply:
> Spec question: why do we want params differences to result in no conversion? The same question applies to optional parameters.

We didn't discuss conversion rules for `params` in particular. However, we discussed that default values should be convertible even if the default value is different but we would introduce a warning wave to indicate this behavior to users.

```csharp
delegate void D(int i = 1);
    
public void M(int i = 2)
{
    // Default value is 2 but warn to user
    D d = M;
}
```

Based on that, we might want to consider the same for `params` scenarios.

Reply:
```
delegate void D(int i = 1);

class C
{
  void M(int i = 2)
  {
    D d = M; // existing code but no diagnostic
  }
}
```

This is existing C# code and it produces no diagnostics ([sharplab](https://sharplab.io/#v2:C4LglgNgNAJiDUAfGBTCKDmBDYKAEAAgCx4AiAFGAHbB5h4C8eAjAJQDcAsAFA8EBMeAMI8A3jzyESAWUo06jPP1YS847pMkB6LWRQAzLAFcItAG5YIR/GADOSvACMjtAO5YATlTzAA9niNbFA9VSVI8GEVpLg08AF8eBO4gA===)).
So I think that the anonymous function conversion should exist regardless of differences in default or `params` differences, just like we already do in method group conversions.

Reply:
Good catch. When implementing `params`, I was following the prior implementation of optional parameters, but seems it was wrong here. The [proposal](https://github.com/dotnet/csharplang/blob/main/proposals/lambda-method-group-defaults.md) has this example:

```cs
void Fun(int i = 4) { ... }

delegate void D1(int i);
D1 d = (int i = 1) => 42; // Allowed WITH warning, since the lambda is now only callable through the delegate
                          // and the default value has effectively been thrown away.

D1 d = Fun; // Allowed WITHOUT warning, since Fun and its default parameter value are still accessible
            // without using the delegate.
```

But it currently results in an error, not a warning. I guess I will fix this in a separate PR and then update this one.

> do we need a spec update for anonymous function conversions

Another good catch. Yes, I will open a PR to the proposal.

Reply:
Actually, I now think this is correct. The examples you provided are still allowed:

```cs
delegate void D(params int[] i);
public void M(int[] i) 
{
    D d = M; // no diagnostic
}
```
```cs
delegate void D(int i = 1);
public void M(int i = 2)
{
    D d = M; // warning CS9502: Parameter 1 has default value '2' in method group and '1' in the target delegate type.
}
```

By the proposal, when using lambdas with optional (or `params`) parameters, the conversion is expected to be an error (it's not an existing code and the optional parameter is useless):

```cs
delegate void D(int i = 1);
D lam = (int i = 2) => { }; // error CS9501: Parameter 1 has default value '2' in lambda and '1' in the target delegate type.
```

Only the example I linked from the proposal is a special case where the lambda has a default parameter value and the delegate doesn't have one:

```cs
delegate void D1(int i);
D1 d = (int i = 1) => 42; // Allowed WITH warning, since the lambda is now only callable through the delegate
                          // and the default value has effectively been thrown away.
```

This is currently an error, I will fix that in another PR.

Reply:
I understand that the scenarios being discussed are new (conversions of lambdas with optional or `params` parameter). We're not breaking existing scenarios.
But the new rules seem inconsistent with existing rules.

To be consistent, we would need the conversion with default value difference to succeed (just a warning) and a `params` difference to succeed also (no diagnostic).
If we prefer the inconsistency, I'd like to understand why.

Reply:
Changed conversion warnings to errors for both `params` and optional parameters (c2d308d1d9eaec5684ec04c174095191b99ae340) as proposed in https://github.com/dotnet/csharplang/pull/6648.